lib/nat/service.go

```diff
@@ -299,7 +310,17 @@ func (s *Service) acquireNewLocked(ctx context.Context, mapping *Mapping, nats m
 
 		l.Debugf("Trying to open port %s on %s", mapping, id)
 
-		addrs, err := s.tryNATDevice(ctx, nat, mapping.address.Port, 0, leaseTime)
+		if mapping.ipVersion == IPv4Only && nat.IsIPv6GatewayDevice() {
+			l.Debugf("Skipping pinholing for %s because listener is IPv4-only", nat.ID())
+			continue
+		}
+
+		if mapping.ipVersion == IPv6Only && !nat.IsIPv6GatewayDevice() {
+			l.Debugf("Skipping port mapping for %s because listener is IPv6-only", nat.ID())
```

Comment:
Inconsistent messages (pinholing / port mapping)?

Reply:
That's a deliberate choice. UPnP calls the IPv4 call `AddPortMapping` and the IPv6 call `AddPinhole` and a "port mapping" doesn't really make sense with the IPv6 code since the ability to map an external port to a different internal one was dropped with IPv6.